contracts/libraries/IterableAppendOnlySet.sol

```diff
@@ -0,0 +1,34 @@
+pragma solidity ^0.5.0;
+
+library IterableAppendOnlySet {
+    struct Data {
+        mapping(address => address) nextMap;
+        address last;
+    }
+
+    function insert(Data storage self, address value) public returns (bool) {
+        if (contains(self, value)) {
+            return false;
+        }
+        self.nextMap[self.last] = value;
+        self.last = value;
+        return true;
+    }
+
+    function contains(Data storage self, address value) public view returns (bool) {
+        require(value != address(0), "Inserting address(0) is not supported");
+        return self.nextMap[value] != address(0) || (self.last == value);
```

Comment:
I would have suggested implementing this as a circularly linked list and you wouldn't need this OR. However, after reading this, I would have reverted my suggestion and said that this is really great.